src/platform/tiles.c

```diff
@@ -0,0 +1,714 @@
+    //  -  Textures[1]: tiles are (W+1) x   H   pixels
+    //  -  Textures[2]: tiles are   W   x (H+1) pixels
+    //  -  Textures[3]: tiles are (W+1) x (H+1) pixels
+
+    for (int i = 0; i < numTextures; i++) {
+
+        // choose dimensions
+        int tileWidth = baseTileWidth + (i == 1 || i == 3 ? 1 : 0);
+        int tileHeight = baseTileHeight + (i == 2 || i == 3 ? 1 : 0);
+        int surfaceWidth = 1, surfaceHeight = 1;
+        while (surfaceWidth < tileWidth * TILE_COLS) surfaceWidth *= 2;
+        while (surfaceHeight < tileHeight * TILE_ROWS) surfaceHeight *= 2;
+
+        // downscale the tiles
+        SDL_Surface *surface = SDL_CreateRGBSurfaceWithFormat(0, surfaceWidth, surfaceHeight, 32, SDL_PIXELFORMAT_ARGB8888);
+        if (!surface) sdlfatal(__FILE__, __LINE__);
+        for (int row = 0; row < TILE_ROWS; row++) {
+            for (int column = 0; column < TILE_COLS; column++) {
+                prepareTile(surface, tileWidth, tileHeight, row, column, false);
+            }
+        }
+
+        // convert to texture
+        Textures[i] = SDL_CreateTextureFromSurface(renderer, surface);
+        if (!Textures[i]) sdlfatal(__FILE__, __LINE__);
+        if (SDL_SetTextureBlendMode(Textures[i], SDL_BLENDMODE_BLEND) < 0) sdlfatal(__FILE__, __LINE__);
+        SDL_FreeSurface(surface);
+    }
+}
+
+
+void updateTile(int row, int column, short charIndex,
+    short foreRed, short foreGreen, short foreBlue,
+    short backRed, short backGreen, short backBlue)
+{
+    screenTiles[row][column] = (ScreenTile){
+        .needsRefresh = 1,
+        .charIndex = charIndex,
+        .foreRed   = foreRed,
+        .foreGreen = foreGreen,
+        .foreBlue  = foreBlue,
+        .backRed   = backRed,
+        .backGreen = backGreen,
+        .backBlue  = backBlue
+    };
+}
+
+
+void updateScreen() {
+    if (!Win) return;
+
+    SDL_Renderer *renderer = SDL_GetRenderer(Win);
+    if (!renderer) {
+        renderer = SDL_CreateRenderer(Win, -1, (softwareRendering ? SDL_RENDERER_SOFTWARE : 0));
+        if (!renderer) sdlfatal(__FILE__, __LINE__);
+
+        if (SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_NONE) < 0) sdlfatal(__FILE__, __LINE__);
+
+        // see if we ended up using the software renderer or not
+        SDL_RendererInfo info;
+        if (SDL_GetRendererInfo(renderer, &info) < 0) sdlfatal(__FILE__, __LINE__);
+        softwareRendering = (strcmp(info.name, "software") == 0);
+    }
+
+    int outputWidth, outputHeight;
+    if (SDL_GetRendererOutputSize(renderer, &outputWidth, &outputHeight) < 0) sdlfatal(__FILE__, __LINE__);
+    if (outputWidth == 0 || outputHeight == 0) return;
+
+    loadTiles(renderer, outputWidth, outputHeight);
+
+    if (!softwareRendering) {
+        // black out the frame (double-buffering invalidated it)
+        if (SDL_SetRenderDrawColor(renderer, 0, 0, 0, 0) < 0) sdlfatal(__FILE__, __LINE__);
+        if (SDL_RenderClear(renderer) < 0) sdlfatal(__FILE__, __LINE__);
+    }
```

Comment:
When using SDL's software renderer, either explicitly with  `--no-gpu` or because the machine doesn't offer hardware acceleration, this optimization drops the CPU from 10% down to 2%, so I'd stay it is essential to have it.

This behavior of the software renderer is indeed not specified in the SDL documentation, but it is true, at least in SDL 2.0.12. The software renderer draws onto a single surface (no double-buffering), which is never invalidated nor modified except through the drawing functions. That's why it works, on all platforms. If this undocumented behavior changes in a future version we can adapt, but until then keeping the CPU as low as possible is a must-have on battery-powered devices.

Reply:
Ok great, good point. I agree it makes sense to assume this, as it's exactly how the surface subsystem works.